lib/smooth/core/io/spi/Master.cpp

```diff
@@ -23,63 +23,101 @@ namespace smooth::core::io::spi
+    bool Master::hspi_initialized = false;
+    bool Master::vspi_initialized = false;
+    std::mutex Master::guard{};
+    spi_bus_config_t Master::bus_config{};
+    spi_host_device_t Master::spi_host;
+    SPI_DMA_Channel Master::dma_channel;
+
+    bool Master::initialize(spi_host_device_t host,
+                            SPI_DMA_Channel dma_chl,
+                            gpio_num_t mosi,
+                            gpio_num_t miso,
+                            gpio_num_t clock,
+                            int transfer_size,
+                            gpio_num_t quadwp_io_num,
+                            gpio_num_t quadhd_io_num)
     {
+        std::lock_guard<std::mutex> lock(guard);
+
+        spi_host = host;
+        dma_channel = dma_chl;
+
         bus_config.mosi_io_num = mosi;
         bus_config.miso_io_num = miso;
         bus_config.sclk_io_num = clock;
         bus_config.quadwp_io_num = quadwp_io_num;
         bus_config.quadhd_io_num = quadhd_io_num;
         bus_config.max_transfer_sz = transfer_size;
-    }
 
-    Master::~Master()
-    {
-        deinitialize();
+        spi_bus_initialized = false;
+
+        if (spi_host == VSPI_HOST)
+        {
+            spi_bus_initialized = do_vspi_initialization();
+        }
+
+        if (spi_host == HSPI_HOST)
+        {
+            spi_bus_initialized = do_hspi_initialization();
+        }
+
+        return spi_bus_initialized;
     }
 
-    bool Master::initialize()
+    bool Master::do_vspi_initialization()
     {
-        std::lock_guard<std::mutex> lock(guard);
-        do_initialization();
+        if (!vspi_initialized)
+        {
+            vspi_initialized = spi_bus_initialize(spi_host, &bus_config, dma_channel) == ESP_OK;
 
-        return initialized;
+            if (!vspi_initialized)
+            {
+                Log::error(log_tag, "VSPI_HOST Initialization failed");
+            }
+            else
+            {
+                Log::info(log_tag, "VSPI initialized, Host {}, DMA {}", spi_host, dma_channel);
+            }
+        }
+
+        return vspi_initialized;
     }
 
-    void Master::do_initialization()
+    bool Master::do_hspi_initialization()
     {
-        if (!initialized)
+        if (!hspi_initialized)
         {
-            initialized = spi_bus_initialize(host, &bus_config, dma_channel) == ESP_OK;
+            hspi_initialized = spi_bus_initialize(spi_host, &bus_config, dma_channel) == ESP_OK;
 
-            if (!initialized)
+            if (!hspi_initialized)
             {
-                Log::error(log_tag, "Initialization failed");
+                Log::error(log_tag, "HSPI_HOST Initialization failed");
             }
             else
             {
-                Log::verbose(log_tag, "SPI initialized, Host {}, DMA {}", host, dma_channel);
+                Log::info(log_tag, "HSPI initialized, Host {}, DMA {}", spi_host, dma_channel);
             }
         }
+
+        return hspi_initialized;
```

Comment:
Do I remember correctly that we can only use either VSPI or HSPI, not both?

These two initializers are very similar and I don't see hspi_initialized or vspi_initialized being used other than to determine what to deinitialize. How about making them into a single function and instead passing hspi or vspi to it? For initialization, we only really need to save which bus is being used and if it is initialized or not.

Reply:
No that is not correct both can be used.  I use vspi for lcd display & sdcard and hspi for connecting to SPI-BME280.  So I think we need both.  Plus this is were I will add the counter.

Reply:
I tested newest code on ESP32-DevKit V1 with 2.8" TFT LCD with Touchscreen Breakout Board w/MicroSD Socket - ILI9341 (LCD and SDCard were on the VSPI_HOST bus) and the BME280-spi was placed on the HSPI_HOST bus.  Both spi-bus work.  The LCD display, SDCard and BME280 work without problems.  Also tested deinit the SDCard and the SDCard was unmounted and the initialized_count prevented the VSPI_HOST bus from being freed so the LCD display work.